<a href="https://colab.research.google.com/github/Domegal13/DATA-ANALITICS-PYTHON-TALENTO-TECH-C25261/blob/main/Gallo_Domenico_Comision_C25261_PFI_Data_Analitics_con_Python_Pre_Entrega.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Etapa 1: Recopilación y Preparación de Datos (Clases 1 a 4)

## Importar librerias

In [1]:
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

## ETL (Extract, Transform, Load) 🛠️

## 1.- Cargar los sets de datos como DataFrames.

### 1.1. Extracción de los Sets de Datos

In [38]:
# clientes = '/content/drive/MyDrive/003-CURSOS/001-TALENTO_TECH/001-DATA ANALYTICS CON PYTHON/Gallo Domenico - Comision C25261 - PFI Data Analitics con Python - Pre_Entrega/Dataset-20250929T222907Z-1-001/clientes.csv'
# marketing = '/content/drive/MyDrive/003-CURSOS/001-TALENTO_TECH/001-DATA ANALYTICS CON PYTHON/Gallo Domenico - Comision C25261 - PFI Data Analitics con Python - Pre_Entrega/Dataset-20250929T222907Z-1-001/marketing.csv'
# ventas = '/content/drive/MyDrive/003-CURSOS/001-TALENTO_TECH/001-DATA ANALYTICS CON PYTHON/Gallo Domenico - Comision C25261 - PFI Data Analitics con Python - Pre_Entrega/Dataset-20250929T222907Z-1-001/ventas.csv'

clientes = 'clientes.csv'
marketing = 'marketing.csv'
ventas = 'ventas.csv'

### 1.2. Cargar los DataFrames

In [3]:
df_clientes = pd.read_csv(clientes)
df_marketing = pd.read_csv(marketing)
df_ventas = pd.read_csv(ventas)

#### Información de los df

In [4]:
display(df_clientes.info())
print('-'*50)
display(df_marketing.info())
print('-'*50)
display(df_ventas.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 567 entries, 0 to 566
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   id_cliente  567 non-null    int64  
 1   nombre      567 non-null    object 
 2   edad        567 non-null    int64  
 3   ciudad      567 non-null    object 
 4   ingresos    567 non-null    float64
dtypes: float64(1), int64(2), object(2)
memory usage: 22.3+ KB


None

--------------------------------------------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 90 entries, 0 to 89
Data columns (total 6 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   id_campanha   90 non-null     int64  
 1   producto      90 non-null     object 
 2   canal         90 non-null     object 
 3   costo         90 non-null     float64
 4   fecha_inicio  90 non-null     object 
 5   fecha_fin     90 non-null     object 
dtypes: float64(1), int64(1), object(4)
memory usage: 4.3+ KB


None

--------------------------------------------------
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3035 entries, 0 to 3034
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id_venta     3035 non-null   int64  
 1   producto     3035 non-null   object 
 2   precio       3033 non-null   object 
 3   cantidad     3033 non-null   float64
 4   fecha_venta  3035 non-null   object 
 5   categoria    3035 non-null   object 
dtypes: float64(1), int64(1), object(4)
memory usage: 142.4+ KB


None

In [5]:
df_ventas.sample(5)

id_venta               producto   precio  cantidad fecha_venta  \
2577       587  Elementos de cerámica   $33.98       9.0  05/11/2024   
3031      2276                 Laptop   $85.27       9.0  30/12/2024   
892       1361     Freidora eléctrica  $103.64      10.0  09/04/2024   
630       1772               Heladera   $56.94       2.0  14/03/2024   
1699      1494             Aspiradora   $82.50       6.0  15/07/2024   

              categoria  
2577         Decoración  
3031        Electrónica  
892   Electrodomésticos  
630   Electrodomésticos  
1699  Electrodomésticos

In [6]:
df_ventas.duplicated().sum()

np.int64(35)

### ✅ Conclusión: 1:
* A simple rasgos se puede observar que existen sólo 2 valores Nulos tanto en el campo 'Precio?' y 'Cantidad'.
* El tipo de dato del campo 'precio' es 'object' - Convertitr a 'float'.
* Los valores del campo 'precio' tiene el carater '$' - Hay que eliminarlo para poder convertir a tipo 'float'.
* Convertir 'fecha_venta' en 'datetime'.
* Se observa que existen 35 registros duplicados. - Hay que eliminarlos.

### 1.3. Transformación de los Datos

#### Limpieza de Datos

#### Se liminan los duplicados

In [7]:
df_ventas = df_ventas.drop_duplicates(keep='first').reset_index(drop=True)
df_ventas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3000 entries, 0 to 2999
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id_venta     3000 non-null   int64  
 1   producto     3000 non-null   object 
 2   precio       2998 non-null   object 
 3   cantidad     2998 non-null   float64
 4   fecha_venta  3000 non-null   object 
 5   categoria    3000 non-null   object 
dtypes: float64(1), int64(1), object(4)
memory usage: 140.8+ KB


#### Se eliminan los registros vacios

In [8]:
df_ventas.dropna(inplace=True)
df_ventas.reset_index(drop=True, inplace=True)
df_ventas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2998 entries, 0 to 2997
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id_venta     2998 non-null   int64  
 1   producto     2998 non-null   object 
 2   precio       2998 non-null   object 
 3   cantidad     2998 non-null   float64
 4   fecha_venta  2998 non-null   object 
 5   categoria    2998 non-null   object 
dtypes: float64(1), int64(1), object(4)
memory usage: 140.7+ KB


In [9]:
df_ventas.sample(4)

id_venta                producto  precio  cantidad fecha_venta  \
1286      1721      Freidora eléctrica  $26.06       5.0  26/05/2024   
2866       546                Cortinas  $82.12      10.0  15/12/2024   
1791      2130  Consola de videojuegos  $41.12       6.0  31/07/2024   
2589      2087          Cámara digital  $35.44       3.0  10/11/2024   

              categoria  
1286  Electrodomésticos  
2866         Decoración  
1791        Electrónica  
2589        Electrónica

#### Se convierten los campos  "cantidad" y  "precio"  a int y float

In [10]:
# 'cantidad' a entero
df_ventas['cantidad'] = pd.to_numeric(df_ventas['cantidad'], errors='coerce').fillna(0).astype(int)


In [11]:
# Se eliminan los espacios al principio y al final, el simbolo '$'
df_ventas['precio'] = (
    df_ventas['precio']
    .astype(str)
    .str.replace(',', '.')
    .str.replace('$', '', regex=False)
    .str.strip()
    .pipe(pd.to_numeric, errors='coerce')
)

precio_promedio = df_ventas['precio'].mean().round(2)
df_ventas['precio'].fillna(precio_promedio, inplace=True)

df_ventas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2998 entries, 0 to 2997
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   id_venta     2998 non-null   int64  
 1   producto     2998 non-null   object 
 2   precio       2998 non-null   float64
 3   cantidad     2998 non-null   int64  
 4   fecha_venta  2998 non-null   object 
 5   categoria    2998 non-null   object 
dtypes: float64(1), int64(2), object(3)
memory usage: 140.7+ KB


## 2.- Ventas Mensuales.

#### 2.1. Se crean los campos mes y año

In [12]:
df_ventas['fecha_venta'] = pd.to_datetime(df_ventas['fecha_venta'], format='%d/%m/%Y')
df_ventas['mes'] = df_ventas['fecha_venta'].dt.month
df_ventas['anio'] = df_ventas['fecha_venta'].dt.year

In [13]:
display(df_ventas.sample(5))

id_venta                producto  precio  cantidad fecha_venta  \
776       1261      Freidora eléctrica   84.29        10  2024-03-31   
1869      2798              SmartWatch   56.58        12  2024-08-09   
403       2246                  Laptop   76.02        10  2024-02-17   
796        226                Cortinas   27.05         3  2024-04-03   
505       2110  Consola de videojuegos   81.69         4  2024-02-28   

              categoria  mes  anio  
776   Electrodomésticos    3  2024  
1869        Electrónica    8  2024  
403         Electrónica    2  2024  
796          Decoración    4  2024  
505         Electrónica    2  2024

#### 2.2. se crea el campo total_venta

In [14]:
df_ventas['total_venta'] = df_ventas['precio'] * df_ventas['cantidad']
df_ventas.sample(4)

id_venta               producto  precio  cantidad fecha_venta  \
2464      1354             Aspiradora   55.17         4  2024-10-26   
1408       637  Elementos de cerámica   44.23         4  2024-06-11   
1209      2971    Parlantes Bluetooth   74.59         7  2024-05-17   
974        538      Rincón de plantas  119.89         4  2024-04-22   

              categoria  mes  anio  total_venta  
2464  Electrodomésticos   10  2024       220.68  
1408         Decoración    6  2024       176.92  
1209        Electrónica    5  2024       522.13  
974          Decoración    4  2024       479.56

#### 2.3. Se calculan las Ventas Mensuales

---



In [15]:
ventas_mensuales = df_ventas.groupby("mes")["total_venta"].sum().reset_index()
ventas_mensuales.rename(columns={'total_venta': 'ventas_mensuales'}, inplace=True)
ventas_mensuales

mes  ventas_mensuales
0     1         129604.99
1     2         118672.44
2     3         136779.15
3     4         128430.69
4     5         143727.25
5     6         108480.17
6     7         116229.97
7     8         119680.15
8     9         115787.85
9    10         112117.13
10   11         119951.79
11   12         117631.94

In [16]:
mapear_meses = {
    1: 'Enero',
    2: 'Febrero',
    3: 'Marzo',
    4: 'Abril',
    5: 'Mayo',
    6: 'Junio',
    7: 'Julio',
    8: 'Agosto',
    9: 'Septiembre',
    10: 'Octubre',
    11: 'Noviembre',
    12: 'Diciembre'
}

ventas_mensuales['mes'] = ventas_mensuales['mes'].map(mapear_meses)
ventas_mensuales

mes  ventas_mensuales
0        Enero         129604.99
1      Febrero         118672.44
2        Marzo         136779.15
3        Abril         128430.69
4         Mayo         143727.25
5        Junio         108480.17
6        Julio         116229.97
7       Agosto         119680.15
8   Septiembre         115787.85
9      Octubre         112117.13
10   Noviembre         119951.79
11   Diciembre         117631.94

In [17]:
ventas_mensuales['ventas_mensuales'].describe()

count        12.000000
mean     122257.793333
std       10361.463891
min      108480.170000
25%      116119.440000
50%      119176.295000
75%      128724.265000
max      143727.250000
Name: ventas_mensuales, dtype: float64

In [18]:
mes_menos_ventas = ventas_mensuales.nsmallest(n=1, columns='ventas_mensuales')
mes_mas_ventas = ventas_mensuales.nlargest(n=1, columns='ventas_mensuales')
display("Mes con Menor Ventas")
display(mes_menos_ventas)
display("-"*50)
display('Mes con Mayor Ventas')
display(mes_mas_ventas)


'Mes con Menor Ventas'

mes  ventas_mensuales
5  Junio         108480.17

'--------------------------------------------------'

'Mes con Mayor Ventas'

mes  ventas_mensuales
4  Mayo         143727.25

### Ventas totales por categoria

In [19]:
ventas_por_categoria = df_ventas.groupby('categoria')['total_venta'].sum().sort_values(ascending=False)

print("--- Ventas Totales por Categoría ---")
print(ventas_por_categoria)

--- Ventas Totales por Categoría ---
categoria
Electrodomésticos    505299.63
Electrónica          482577.80
Decoración           479216.09
Name: total_venta, dtype: float64


#### Ordenar por ventas, de mayor a menor (descendente)

In [20]:

df_ranking_ventas = ventas_mensuales.sort_values(by='ventas_mensuales', ascending=False).reset_index(drop=True)

print("--- Ranking por Ventas (Mayor a Menor) ---")
print(df_ranking_ventas)

--- Ranking por Ventas (Mayor a Menor) ---
           mes  ventas_mensuales
0         Mayo         143727.25
1        Marzo         136779.15
2        Enero         129604.99
3        Abril         128430.69
4    Noviembre         119951.79
5       Agosto         119680.15
6      Febrero         118672.44
7    Diciembre         117631.94
8        Julio         116229.97
9   Septiembre         115787.85
10     Octubre         112117.13
11       Junio         108480.17


### Meses con Mayores Ventas

In [21]:
top_ventas_por_mes = df_ranking_ventas.nlargest(n=3, columns='ventas_mensuales').reset_index(drop=True)
print("--- Top 3 Meses con Mayor Venta ---")
display(top_ventas_por_mes)

--- Top 3 Meses con Mayor Venta ---


mes  ventas_mensuales
0   Mayo         143727.25
1  Marzo         136779.15
2  Enero         129604.99

### Ventas Totales por Producto

In [22]:
# ventas_por_producto = df_ventas.groupby('producto')['total_venta'].sum().sort_values(ascending=False)
ventas_por_producto =(df_ventas.groupby('producto', as_index=False)
                                .agg(unidades_vendidas=('cantidad', 'sum'),
                                     ventas_totales=('total_venta', 'sum'))
                                .sort_values(by='ventas_totales', ascending=False)
                                .reset_index(drop=True)).round(2)

print('-'*100)
print("--- Ventas Totales por Producto ---")
print('-'*100)
print(ventas_por_producto.head())

----------------------------------------------------------------------------------------------------
--- Ventas Totales por Producto ---
----------------------------------------------------------------------------------------------------
            producto  unidades_vendidas  ventas_totales
0    Lámpara de mesa               1112        82276.38
1        Auriculares                958        74175.58
2         Microondas                912        72562.89
3           Cafetera                765        59607.31
4  Cuadro decorativo                726        54297.60


### Productos con Mayores Ventas

In [23]:
n=3
top_ventas_por_producto = ventas_por_producto.nlargest(n, 'ventas_totales').reset_index(drop=True)
print('-'*100)
print("--- Top 3 Productos con Mayor Venta ---")
print('-'*100)
display(top_ventas_por_producto)

----------------------------------------------------------------------------------------------------
--- Top 3 Productos con Mayor Venta ---
----------------------------------------------------------------------------------------------------


producto  unidades_vendidas  ventas_totales
0  Lámpara de mesa               1112        82276.38
1      Auriculares                958        74175.58
2       Microondas                912        72562.89

In [24]:
display(ventas_por_producto.describe())

unidades_vendidas  ventas_totales
count          30.000000       30.000000
mean          649.833333    48903.117333
std           169.020522    13265.561715
min           160.000000    11128.800000
25%           623.750000    45601.920000
50%           635.500000    48140.245000
75%           672.000000    51092.960000
max          1112.000000    82276.380000

## Combinar los sets de datos de ventas y marketing


In [25]:
display('Ventas')
display(df_ventas.sample(5))
display('Marketing')
display(df_marketing.sample(5))

'Ventas'

id_venta                producto  precio  cantidad fecha_venta  \
1470      2455             Auriculares   91.06         3  2024-06-19   
1284      1025                Lavadora  119.78         9  2024-05-26   
2230        93       Espejo decorativo   44.57         9  2024-09-25   
341       2668              SmartWatch   97.01         4  2024-02-09   
1815      2880  Consola de videojuegos   79.76        12  2024-08-02   

              categoria  mes  anio  total_venta  
1470        Electrónica    6  2024       273.18  
1284  Electrodomésticos    5  2024      1078.02  
2230         Decoración    9  2024       401.13  
341         Electrónica    2  2024       388.04  
1815        Electrónica    8  2024       957.12

'Marketing'

id_campanha             producto  canal  costo fecha_inicio   fecha_fin
1            12               Tablet   RRSS   3.40   26/03/2024  13/05/2024
9            84  Parlantes Bluetooth     TV   4.42   17/04/2024  05/05/2024
84           28             Alfombra   RRSS   5.82   27/11/2024    8/1/2025
60           13             Cortinas   RRSS   5.22   16/09/2024  15/10/2024
13           34             Heladera  Email   4.20   06/05/2024  05/07/2024

In [26]:
df_marketing['producto'].value_counts().sample(3).reset_index()

producto  count
0  Smartphone      3
1    Batidora      3
2  Aspiradora      3

### Agrupación de Marketing por Producto

In [27]:
marketin_producto = (df_marketing.groupby('producto')
                                 .agg(campañas=('producto', 'count'),
                                      inversion_total = ('costo', 'sum'))
.sort_values('inversion_total', ascending=False)
.reset_index()).round(2)

print('-'*100)
print("--- Costo de Marketin por Producto ---")
print('-'*100)
print(marketin_producto.head())

----------------------------------------------------------------------------------------------------
--- Costo de Marketin por Producto ---
----------------------------------------------------------------------------------------------------
                producto  campañas  inversion_total
0               Cafetera         3            18.16
1               Alfombra         3            17.46
2               Cortinas         3            17.36
3  Elementos de cerámica         3            16.87
4             Aspiradora         3            16.46


### Se Combinan los DF "ventas_por_productos" y "marketing_producto"

In [28]:
productos_vta_mkt = pd.merge(marketin_producto, ventas_por_producto, on='producto', how='inner')
productos_vta_mkt.head()

producto  campañas  inversion_total  unidades_vendidas  \
0               Cafetera         3            18.16                765   
1               Alfombra         3            17.46                615   
2               Cortinas         3            17.36                610   
3  Elementos de cerámica         3            16.87                636   
4             Aspiradora         3            16.46                651   

   ventas_totales  
0        59607.31  
1        44773.06  
2        44865.03  
3        45411.09  
4        50085.86

### Indicadores Derivados

In [29]:
productos_vta_mkt['ticket_promedio'] = (productos_vta_mkt['ventas_totales'] / productos_vta_mkt['unidades_vendidas']).round(2)
productos_vta_mkt.head(8)

producto  campañas  inversion_total  unidades_vendidas  \
0               Cafetera         3            18.16                765   
1               Alfombra         3            17.46                615   
2               Cortinas         3            17.36                610   
3  Elementos de cerámica         3            16.87                636   
4             Aspiradora         3            16.46                651   
5      Rincón de plantas         3            16.45                691   
6         Cámara digital         3            16.40                638   
7        Lámpara de mesa         3            15.93               1112   

   ventas_totales  ticket_promedio  
0        59607.31            77.92  
1        44773.06            72.80  
2        44865.03            73.55  
3        45411.09            71.40  
4        50085.86            76.94  
5        50456.45            73.02  
6        45217.96            70.87  
7        82276.38            73.99

In [30]:
productos_vta_mkt['ROAS'] = (productos_vta_mkt['ventas_totales']/productos_vta_mkt['inversion_total']).round(2)
productos_vta_mkt.head(8)

producto  campañas  inversion_total  unidades_vendidas  \
0               Cafetera         3            18.16                765   
1               Alfombra         3            17.46                615   
2               Cortinas         3            17.36                610   
3  Elementos de cerámica         3            16.87                636   
4             Aspiradora         3            16.46                651   
5      Rincón de plantas         3            16.45                691   
6         Cámara digital         3            16.40                638   
7        Lámpara de mesa         3            15.93               1112   

   ventas_totales  ticket_promedio     ROAS  
0        59607.31            77.92  3282.34  
1        44773.06            72.80  2564.32  
2        44865.03            73.55  2584.39  
3        45411.09            71.40  2691.83  
4        50085.86            76.94  3042.88  
5        50456.45            73.02  3067.26  
6        45217.96            70.87  2757.19  
7        82276.38            73.99  5164.87

In [31]:
productos_vta_mkt = productos_vta_mkt.sort_values('ROAS', ascending=False).reset_index(drop=True)
# productos_vta_mkt.head()
# display(productos_vta_mkt['ROAS'].max())
# display(productos_vta_mkt['ROAS'].min())

###  TOP 3 Productos con Mayor Rendimiento

In [32]:
n = 3

In [33]:
productos_con_mayor_rendimiento = productos_vta_mkt.nlargest(n=n, columns='ROAS').reset_index(drop=True)
productos_con_mayor_rendimiento

producto  campañas  inversion_total  unidades_vendidas  \
0  Lámpara de mesa         3            15.93               1112   
1       Microondas         3            14.19                912   
2      Auriculares         3            15.24                958   

   ventas_totales  ticket_promedio     ROAS  
0        82276.38            73.99  5164.87  
1        72562.89            79.56  5113.66  
2        74175.58            77.43  4867.16

###  TOP 3 Productos con Mayores Ventas

In [34]:
productos_con_mayores_ventas = productos_vta_mkt.nlargest(n, columns='ventas_totales').reset_index(drop=True)
productos_con_mayores_ventas

producto  campañas  inversion_total  unidades_vendidas  \
0  Lámpara de mesa         3            15.93               1112   
1      Auriculares         3            15.24                958   
2       Microondas         3            14.19                912   

   ventas_totales  ticket_promedio     ROAS  
0        82276.38            73.99  5164.87  
1        74175.58            77.43  4867.16  
2        72562.89            79.56  5113.66

In [35]:
display('Producto con Mayor Roas')
display(productos_vta_mkt.loc[productos_vta_mkt['ROAS'] == productos_vta_mkt['ROAS'].max()])
display('-'*100)
display('roducto con Menor Roas')
display(productos_vta_mkt.loc[productos_vta_mkt['ROAS'] == productos_vta_mkt['ROAS'].min()])

'Producto con Mayor Roas'

producto  campañas  inversion_total  unidades_vendidas  \
0  Lámpara de mesa         3            15.93               1112   

   ventas_totales  ticket_promedio     ROAS  
0        82276.38            73.99  5164.87

'----------------------------------------------------------------------------------------------------'

'roducto con Menor Roas'

producto  campañas  inversion_total  unidades_vendidas  ventas_totales  \
29  Candelabro         3            14.65                160         11128.8   

    ticket_promedio    ROAS  
29            69.55  759.65

In [36]:
# productos_vta_mkt.loc[productos_vta_mkt['ROAS'] == productos_vta_mkt['ROAS'].max()]
# productos_vta_mkt.loc[productos_vta_mkt['ROAS'] == productos_vta_mkt['ROAS'].min()]

### ✅ Conclusión:

- Todos los productos tienen un ROAS muy alto, lo que indica que las campañas son altamente rentables.
- El producto con mejor rendimiento es la "Lámpara de mesa" (ROAS=5164.87).
- El “Candelabro” tiene un (ROAS=759.65), aunque parece Alto, comparado con los demás productos está rindiemdo mucho menos.


### Resultado:

- Vemos **una fila por producto**, con:
  - `ventas_total`: suma global de ingresos (VENTAS).
  - `unidades_vendidas`: total de unidades vendidas.
  - `inversion_total`: suma global de costos de marketing.
  - `campañas`: cuántas campañas hubo para ese producto.
  - `ticket_medio`: promedio de venta por unidad.
  - `ROAS`: retorno sobre gasto publicitario (ventas / inversión).



## Exportar a .csv

In [37]:
productos_vta_mkt.to_csv('productos_ventas_marketing.csv', index=False)